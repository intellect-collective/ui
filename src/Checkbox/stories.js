import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Checkbox from './index';

storiesOf('Checkbox', module)
    .add('default', () => (<Checkbox name="test-check-1" value="hotdogs" onClick={ action('onClick') } />));
