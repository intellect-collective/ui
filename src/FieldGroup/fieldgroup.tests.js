import React from 'react';
import { FieldGroup, Text } from '../..';

describe('Form Components - FieldGroup', () => {
    it('Can render with string wrapper', () => {
        const wrapper = mount(
            <FieldGroup name="test" wrapper="div">
                <Text />
            </FieldGroup>
        );
        expect(wrapper.find('div').length).toEqual(1);
        expect(wrapper.find('div').props()).toMatchObject({
            role: 'group'
        });
    });
});
