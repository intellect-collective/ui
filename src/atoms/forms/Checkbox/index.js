import React from 'react';
import changeable from '../_decorators/changeable';
import formContext from '../_decorators/formContext';
import groupContext from '../_decorators/groupContext';
import {
    checkedGroupTransformer,
    checkedFormTransformer
} from '../_decorators/_utils';

const Checkbox = (props) => (
    <input { ...props } type="checkbox" />
);
Checkbox.displayName = 'Checkbox';

export default groupContext(
    formContext(changeable(Checkbox), checkedFormTransformer),
    checkedGroupTransformer
);
