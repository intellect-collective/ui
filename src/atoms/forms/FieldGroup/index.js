import React from 'react';
import PropTypes from 'prop-types';
import { groupContext, groupValue } from '../prop-types';

const getValue = (value) => {
    if (value === null || value === undefined) {
        return [];
    }
    if (Array.isArray(value)) {
        return [...value];
    }
    return [value];
};

export default class FieldGroup extends React.Component {
    static propTypes = {
        name: PropTypes.string,
        multiple: PropTypes.bool,
        value: groupValue,
        onChange: PropTypes.func,
        children: PropTypes.node.isRequired,
        component: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.func,
            PropTypes.object
        ])
    };

    static defaultProps = {
        component: 'div'
    };

    static childContextTypes = {
        group: groupContext
    };

    constructor() {
        super();
        this.onChange = this.onChange.bind(this);
    }

    getChildContext() {
        const { name, value } = this.props;
        return {
            group: {
                name,
                value,
                onChange: this.onChange
            }
        };
    }

    onChange(name, value, ev) {
        const { checked } = ev.target;

        // If singular
        if (!this.props.multiple) {
            this.props.onChange(ev);
            return;
        }

        // Otherwise multiple values allowed
        const val = getValue(this.props.value);
        if (checked) {
            val.push(value);
        } else {
            let index = val.indexOf(value);
            while (index !== -1) {
                val.splice(index, 1);
                index = val.indexOf(value);
            }
        }
        this.props.onChange(ev);
    }

    render() {
        const {
            name, value, onChange, component: Component, children, ...rest
        } = this.props;
        return (
            <Component role="group" { ...rest }>
                { children }
            </Component>
        );
    }
}
