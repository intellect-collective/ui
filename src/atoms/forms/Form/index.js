import React from 'react';
import PropTypes from 'prop-types';

export const FormContext = React.createContext(null);

export default class Form extends React.Component {
    static propTypes = {
        action: PropTypes.string.isRequired,
        method: PropTypes.oneOf(['get', 'post', 'put']),
        overrideProperty: PropTypes.string,
        encType: PropTypes.oneOf(['text', 'multipart']),
        onChange: PropTypes.func,
        onSubmit: PropTypes.func,
        children: PropTypes.any,
        data: PropTypes.object
    };

    static defaultProps = {
        method: 'post',
        overrideProperty: '_method',
        data: {},
        onChange: () => {},
        onSubmit: () => {}
    };

    constructor(props) {
        super(props);
        this.getValue = this.getValue.bind(this);
        this.setValue = this.setValue.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.getEncType = this.getEncType.bind(this);
        this.getFormMethod = this.getFormMethod.bind(this);
        this.getMethodOverrideElement = this.getMethodOverrideElement.bind(this);
    }


    getValue(name) {
        if (this.props.data) {
            return this.props.data[name];
        }
        return undefined;
    }

    setValue(ev) {
        this.props.onChange(ev);
    }

    /**
     * Calls the onSubmit prop function, if provided, when the form is
     * submitted.
     * @param {Object} e - The React event
     */
    onSubmit(e) {
        if (this.props.onSubmit) {
            this.props.onSubmit(this.props.data, e);
        }
    }

    /**
     * Determines the encoding type of the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The mime type to specify in th enctype attribute
     */
    getEncType() {
        if (this.props.encType === 'text') {
            return 'text/plain';
        }
        if (this.props.encType === 'multipart') {
            return 'multipart/form-data';
        }
        return this.props.encType;
    }

    /**
     * Determines the HTTP method to use when submitting the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The http method to use
     */
    getFormMethod() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return this.props.method;
        }
        return 'post';
    }

    /**
     * Produces a hidden element which provides the opportunity to emulate
     * http methods other than GET and POST. Requires that the server receiving
     * the request support method overriding.
     * @returns {Object} - A React element
     */
    getMethodOverrideElement() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return null;
        }
        return (
            <input type="hidden"
                    name={ this.props.overrideProperty || '_method' }
                    value={ this.props.method } />
        );
    }

    /**
     * Builds the props that are to be passed through to the form element. We
     * need to get rid of some props that aren't supported by the form element,
     * so we take that opportunity here.
     * @returns {Object} - The final props object
     */
    getProps() {
        const props = {
            ...this.props,
            encType: this.getEncType(),
            method: this.getFormMethod(),
            onSubmit: this.onSubmit
        };
        delete props.overrideProperty;
        return props;
    }

    /**
     * Perform render
     * @returns {Object} - A React element
     */
    render() {
        const {
            children,
            onChange,
            onSubmit,
            ...props
        } = this.getProps();
        const ctx = {
            getValue: this.getValue,
            setValue: this.setValue
        };
        return (
            <form { ...props }>
                <FormContext.Provider value={ ctx }>
                    { this.getMethodOverrideElement() }
                    { this.props.children }
                </FormContext.Provider>
            </form>
        );
    }
}
