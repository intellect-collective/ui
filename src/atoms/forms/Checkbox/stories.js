import React from 'react';
import { storiesOf } from '@storybook/react';
import Checkbox from './index';

storiesOf('Atoms / Forms /Checkbox', module)
    .add('default', () => (<Checkbox name="test-check-1" value="hotdogs" />));
