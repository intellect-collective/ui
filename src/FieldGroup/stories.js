import React from 'react';
import { storiesOf } from '@storybook/react';
import { withState } from '@dump247/storybook-state';
import FieldGroup from './index';
import Checkbox from '../Checkbox';
import Radio from '../Radio';

storiesOf('FieldGroup', module)
    .add('radio buttons', withState({ initialState: {} })(({ store }) => (
        <FieldGroup name="test" value={ store.state.test } onChange={ (ev) => store.set({ [ev.target.name]: ev.target.value }) }>
            First
            {' '}
            <Radio value="first" />
            <br />
            Second
            {' '}
            <Radio value="second" />
            <br />
            Third
            {' '}
            <Radio value="third" />
        </FieldGroup>
    )))
    .add('checkboxes', withState({})(({ store }) => (
        <FieldGroup name="test" value={ store.state.test } multiple onChange={ store.set }>
            First
            {' '}
            <Checkbox value="first" />
            <br />
            Second
            {' '}
            <Checkbox value="second" />
            <br />
            Third
            {' '}
            <Checkbox value="third" />
        </FieldGroup>
    )));
