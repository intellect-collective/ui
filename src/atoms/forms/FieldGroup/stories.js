import React from 'react';
import { storiesOf } from '@storybook/react';
import FieldGroup from './index';
import Checkbox from '../Checkbox';
import Radio from '../Radio';
import stateful from '../../../../.storybook/decorators/stateful';

storiesOf('Atoms / Forms / FieldGroup', module)
    .addDecorator(stateful())
    .add('radio buttons', () => (onChange, state) => (
        <FieldGroup name="test" value={ state.test } onChange={ onChange }>
            First
            {' '}
            <Radio value="first" />
            <br />
            Second
            {' '}
            <Radio value="second" />
            <br />
            Third
            {' '}
            <Radio value="third" />
        </FieldGroup>
    ))
    .add('checkboxes', () => (onChange, state) => (
        <FieldGroup name="test" value={ state.test } multiple onChange={ onChange }>
            First
            {' '}
            <Checkbox value="first" />
            <br />
            Second
            {' '}
            <Checkbox value="second" />
            <br />
            Third
            {' '}
            <Checkbox value="third" />
        </FieldGroup>
    ));
