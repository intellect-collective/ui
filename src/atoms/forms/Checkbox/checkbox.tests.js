import React from 'react';
import Checkbox from '.';
import Form from '../Form';
import FieldGroup from '../FieldGroup';
import {
    basicChangeableTests
} from '../../../../test/utils';

const validProps = {
    name: 'test',
    value: 'fire',
    onChange: () => {}
};

const expectedProps = {
    name: 'test',
    value: 'fire',
    onChange: expect.any(Function)
};

describe('Checkbox', () => {
    basicChangeableTests(Checkbox, 'input', validProps, expectedProps);

    describe('Form Context Usage', () => {
        it('Gets checked state from form', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <Checkbox name="test" value="sausages" />
                </Form>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(true);
        });

        it('Prefers local attributes over form attributes', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <Checkbox name="test" value="sausages" checked={ false } />
                </Form>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Remains unchecked if form field has incorrect value', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'hotdogs' }}>
                    <Checkbox name="test" value="sausages" />
                </Form>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Can check multiple', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: ['hotdogs', 'sausages'] }}>
                    <Checkbox name="test" value="hotdogs" />
                    <Checkbox name="test" value="sausages" />
                    <Checkbox name="test" value="burgers" />
                </Form>
            );
            expect(wrapper.find('input[value="hotdogs"]').prop('checked')).toEqual(true);
            expect(wrapper.find('input[value="sausages"]').prop('checked')).toEqual(true);
            expect(wrapper.find('input[value="burgers"]').prop('checked')).toEqual(false);
        });
    });

    describe('FieldGroup Context Usage', () => {
        it('Gets name and checked state from group', () => {
            const wrapper = mount(
                <FieldGroup name="test" value="sausages">
                    <Checkbox value="sausages" />
                </FieldGroup>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test');
            expect(wrapper.find('input').prop('checked')).toEqual(true);
        });

        it('Prefers local attributes over group attributes', () => {
            const wrapper = mount(
                <FieldGroup name="test" value="sausages">
                    <Checkbox value="sausages" checked={ false } />
                </FieldGroup>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Prefers local name over group name', () => {
            const wrapper = mount(
                <FieldGroup name="test" value="sausages">
                    <Checkbox name="test2" value="sausages" />
                </FieldGroup>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test2');
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Remains unchecked if group has incorrect value', () => {
            const wrapper = mount(
                <FieldGroup name="test" value="hotdogs">
                    <Checkbox value="sausages" />
                </FieldGroup>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Can check multiple', () => {
            const wrapper = mount(
                <FieldGroup name="test" value={ ['hotdogs', 'sausages'] }>
                    <Checkbox value="hotdogs" />
                    <Checkbox value="sausages" />
                    <Checkbox value="burgers" />
                </FieldGroup>
            );
            expect(wrapper.find('input[value="hotdogs"]').prop('checked')).toEqual(true);
            expect(wrapper.find('input[value="sausages"]').prop('checked')).toEqual(true);
            expect(wrapper.find('input[value="burgers"]').prop('checked')).toEqual(false);
        });
    });

    describe('Form + FieldGroup Context Usage', () => {
        it('Gets state from form', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <FieldGroup name="test">
                        <Checkbox value="sausages" />
                    </FieldGroup>
                </Form>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test');
            expect(wrapper.find('input').prop('checked')).toEqual(true);
        });

        it('Prefers state from group', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <FieldGroup name="test" value="hotdogs">
                        <Checkbox value="sausages" />
                    </FieldGroup>
                </Form>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test');
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Prefers state from field', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <FieldGroup name="test" value="hotdogs">
                        <Checkbox value="sausages" checked={ false } />
                    </FieldGroup>
                </Form>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test');
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });
    });
});
