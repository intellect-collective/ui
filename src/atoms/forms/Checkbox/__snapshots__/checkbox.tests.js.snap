// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Should accept valid props 1`] = `
<GroupContext
  name="test"
  onChange={[Function]}
  value="fire"
>
  <FormContext
    name="test"
    onChange={[Function]}
    value="fire"
  >
    <Changeable
      checked={false}
      name="test"
      onChange={[Function]}
      value="fire"
    >
      <Checkbox
        checked={false}
        name="test"
        onChange={[Function]}
        value="fire"
      >
        <input
          checked={false}
          name="test"
          onChange={[Function]}
          type="checkbox"
          value="fire"
        />
      </Checkbox>
    </Changeable>
  </FormContext>
</GroupContext>
`;
